src/transactions/TransactionUtils.cpp

```diff
@@ -0,0 +1,199 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/LedgerEntryReference.h"
+#include "ledger/LedgerHeaderReference.h"
+#include "ledger/LedgerState.h"
+#include "transactions/TransactionUtils.h"
+#include "util/types.h"
+
+#include "ledger/AccountReference.h"
+#include "ledger/TrustLineReference.h"
+
+namespace stellar
+{
+using xdr::operator==;
+
+uint32_t
+getCurrentTxFee(LedgerStateRoot& lsr)
+{
+    LedgerState ls(lsr);
+    return getCurrentTxFee(ls.loadHeader());
+}
+
+uint32_t
+getCurrentTxFee(std::shared_ptr<LedgerHeaderReference> header)
+{
+    return header->header().baseFee;
+}
+
+uint32_t
+getCurrentMaxTxSetSize(LedgerStateRoot& lsr)
+{
+    LedgerState ls(lsr);
+    return getCurrentMaxTxSetSize(ls.loadHeader());
+}
+
+uint32_t
+getCurrentMaxTxSetSize(std::shared_ptr<LedgerHeaderReference> header)
+{
+    return header->header().maxTxSetSize;
+}
+
+uint32_t
+getCurrentLedgerNum(LedgerStateRoot& lsr)
+{
+    LedgerState ls(lsr);
+    return getCurrentLedgerNum(ls.loadHeader());
+}
+
+uint32_t
+getCurrentLedgerNum(std::shared_ptr<LedgerHeaderReference> header)
+{
+    return header->header().ledgerSeq;
+}
+
+uint64_t
+getCurrentCloseTime(LedgerStateRoot& lsr)
+{
+    LedgerState ls(lsr);
+    return getCurrentCloseTime(ls.loadHeader());
+}
+
+uint64_t
+getCurrentCloseTime(std::shared_ptr<LedgerHeaderReference> header)
+{
+    return header->header().scpValue.closeTime;
+}
+
+int64_t
+getCurrentMinBalance(LedgerStateRoot& lsr, uint32_t ownerCount)
+{
+    LedgerState ls(lsr);
+    return getCurrentMinBalance(ls.loadHeader(), ownerCount);
+}
+
+int64_t
+getCurrentMinBalance(std::shared_ptr<LedgerHeaderReference> header, uint32_t ownerCount)
+{
+    if (header->header().ledgerVersion <= 8)
+    {
+        return (2 + ownerCount) * header->header().baseReserve;
+    }
+    else
+    {
+        return (2 + ownerCount) * int64_t(header->header().baseReserve);
+    }
+}
+
+uint32_t
+getCurrentLedgerVersion(LedgerStateRoot& lsr)
+{
+    LedgerState ls(lsr);
+    return getCurrentLedgerVersion(ls.loadHeader());
+}
+
+uint32_t
+getCurrentLedgerVersion(std::shared_ptr<LedgerHeaderReference> header)
+{
+    return header->header().ledgerVersion;
+}
+
+void
+generateOfferID(std::shared_ptr<LedgerEntryReference> offer,
+                std::shared_ptr<LedgerHeaderReference> header)
+{
+    offer->entry()->data.offer().offerID = ++header->header().idPool;
+}
+
+uint64_t
+getStartingSequenceNumber(std::shared_ptr<LedgerHeaderReference> header)
```

Comment:
I think we are overusing shared_ptr all over the code.
For example here a normal (or const) reference would do better:

LedgerHeaderReference const& header

It would not require doing copy of shared_ptr and doing atomic inc/dec of ref-counter. It also suggest that LedgerHeaderReference will not be changes inside this method (which is different in generateOfferID, where non-const referece to offer and header should be passed).